pkg/tracing/tracer.go

```diff
@@ -0,0 +1,272 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go/sync2"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"github.com/pingcap/dm/dm/pb"
+	"github.com/pingcap/dm/pkg/log"
+)
+
+var (
+	uploadInterval = 1 * time.Minute
+)
+
+// Tracer is a tracing service provider
+type Tracer struct {
+	sync.Mutex
+	wg     sync.WaitGroup
+	closed sync2.AtomicBool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	cfg Config
+	cli pb.TracerClient
+
+	tso   *tsoGenerator
+	idGen *IDGenerator
+
+	jobs       map[EventType]chan *Job
+	jobsClosed sync2.AtomicBool
+	rpcWg      sync.WaitGroup
+}
+
+// NewTracer creates a new Tracer
+func NewTracer(cfg Config) *Tracer {
+	t := Tracer{
+		cfg:   cfg,
+		tso:   newTsoGenerator(),
+		idGen: NewIDGen(),
+	}
+	t.jobsClosed.Set(true)
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	return &t
+}
+
+// Enable returns whether tracing is enabled
+func (t *Tracer) Enable() bool {
+	return t.cfg.Enable
+}
+
+// Start starts tracing service
+func (t *Tracer) Start() {
+	if !t.Enable() {
+		return
+	}
+	conn, err := grpc.Dial(t.cfg.TracerAddr, grpc.WithInsecure(), grpc.WithBackoffMaxDelay(3*time.Second))
+	if err != nil {
+		log.Errorf("[tracer] grpc dial error: %s", errors.ErrorStack(err))
+		return
+	}
+	t.cli = pb.NewTracerClient(conn)
+	t.closed.Set(false)
+	t.newJobChans()
+
+	t.wg.Add(1)
+	go func() {
+		ctx2, _ := context.WithCancel(t.ctx)
+		t.tsoProcessor(ctx2)
+		// cancel()
+	}()
+
+	for _, ch := range t.jobs {
+		t.wg.Add(1)
+		go func(c <-chan *Job) {
+			ctx2, _ := context.WithCancel(t.ctx)
+			t.jobProcessor(ctx2, c)
+			// cancel()
+		}(ch)
+	}
+}
+
+// Stop stops tracer
+func (t *Tracer) Stop() {
+	if !t.Enable() {
+		return
+	}
+	t.Lock()
+	defer t.Unlock()
+	if t.closed.Get() {
+		return
+	}
+	if t.cancel != nil {
+		t.cancel()
+	}
+	t.closed.Set(true)
+	t.wg.Wait()
+}
+
+func (t *Tracer) tsoProcessor(ctx context.Context) {
+	defer t.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(1 * time.Minute):
+			err := t.syncTS()
+			if err != nil {
+				log.Errorf("[tracer] sync timestamp error: %s", errors.ErrorStack(err))
+			}
+		}
+	}
+}
+
+// Assume we send GetTSO with local timestamp t1, timestamp gap between local
+// and remote tso service is tx, remote tso service get GetTSO at timestamp t2
+// and send response back as soon as possible. We get RPC response at local
+// timestamp t3. then we have t1 + tx + ttl = t2; t2 - tx + ttl = t3
+// so the real sending RPC request timestamp is t1 + tx = t2 + t1/2 - t3/2
+func (t *Tracer) syncTS() error {
+	t.rpcWg.Add(1)
```

Comment:
where to use `t.rpcWg.Wait()`?

Reply:
Current it is called in unit test only. in which scenario we must wait rpc call returned.